src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c

```diff
@@ -928,19 +896,335 @@ int xocl_cleanup_mem(struct xocl_drm *drm_p)
+			}
+
+                        xocl_info(drm_p->ddev->dev, "Taking down DDR : %d", i);
+                }
+
+		vfree(xocl_mm->mm_usage_stat);
+		xocl_mm->mm_usage_stat = NULL;
+		
+		if (xocl_mm->bo_usage_stat) {
+			vfree(xocl_mm->bo_usage_stat);
+			xocl_mm->bo_usage_stat = NULL;
+		}
+        }
+        XOCL_PUT_MEM_TOPOLOGY(drm_p->xdev);
+
+	/* Now cleanup the P2P memory */
+        xocl_p2p_mem_cleanup(drm_p->xdev);
+
+        /* cleanup the memory manager */
+        if (xocl_mm && xocl_mm->mm) {
+                drm_mm_takedown(xocl_mm->mm);
+		vfree(xocl_mm->mm);
+		vfree(xocl_mm);
+		drm_p->xocl_mm = NULL;		
+	}
+		
+	mutex_unlock(&drm_p->mm_lock);
+
+        return 0;
+}
+
+static int xocl_init_memory_manager(struct xocl_drm *drm_p)
+{
+	struct mem_topology *topo = NULL;
+	struct mem_data *mem_data = NULL;
+        uint64_t mm_start_addr = 0;
+        uint64_t mm_end_addr = 0;
+	size_t ddr_bank_size = 0;
+	int err = 0;
+	int i = 0;
+
+	/* Memory manager initialize should be done only once */
+        if (drm_p->xocl_mm_done)
+                return 0; // Memory manager is already initialized
+
+	mutex_lock(&drm_p->mm_lock);
+        err = xocl_p2p_mem_init(drm_p->xdev);
+        if (err && err != -ENODEV) {
+		mutex_unlock(&drm_p->mm_lock);
+                xocl_err(drm_p->ddev->dev,
+                        "init p2p mem failed, err %d", err);
+                return err;
+        }
+        err = 0;
+
+	err = XOCL_GET_MEM_TOPOLOGY(drm_p->xdev, topo);
+	if (err) {
+		mutex_unlock(&drm_p->mm_lock);
+		return err;
+	}
+
+	/* Initialize with max and min possible value */
+        mm_start_addr = 0xffffFFFFffffFFFF;
+        mm_end_addr = 0;
+	drm_p->cma_bank_idx = -1;
+
+        /* Initialize all the banks and their sizes */
+        /* Currently only fixed sizes are supported */
+        for (i = 0; i < topo->m_count; i++) {
+                mem_data = &topo->m_mem_data[i];
+                ddr_bank_size = mem_data->m_size * 1024;
+
+		if (XOCL_IS_STREAM(topo, i))
+                        continue;
+
+                if (XOCL_IS_PS_KERNEL_MEM(topo, i))
+                        continue;
+
+                if (!is_mem_region_valid(drm_p, mem_data))
+                        continue;
+
+                /* Update the start and end address for the memory manager */
+                if (mem_data->m_base_address < mm_start_addr)
+                        mm_start_addr = mem_data->m_base_address;
+                if ((mem_data->m_base_address + ddr_bank_size) > mm_end_addr)
+                        mm_end_addr = mem_data->m_base_address + ddr_bank_size;
+
+		if (convert_mem_tag(mem_data->m_tag) == MEM_TAG_HOST) {
+			drm_p->cma_bank_idx = i;
+			err = xocl_set_cma_bank(drm_p, mem_data->m_base_address, ddr_bank_size);
+			if (err) {
+				xocl_err(drm_p->ddev->dev, 
+					"Run host_mem to setup host memory access, request 0x%lx bytes",
+					ddr_bank_size);
+				goto error;
+			}
+		}
+	}
+
+	if (drm_p->xocl_mm == NULL) {
```

Comment:
We need to check here. Please check the description of this PR. I have updated the logic of the current memory manager for multislot support.

Reply:
Then, could you do the check earlier so that you don't waste time figuring out mm_start_addr and mm_end_addr only to find that you don't need them here?